File: ruby/say/say.rb
class Say
  NUMBERS = %w(zero one two three four five six
    seven eight nine ten eleven twelve thirteen fourteen fifteen
    sixteen seventeen eighteen nineteen)

  TENS = %w(twenty thirty forty fifty sixty seventy eighty ninety)

  def initialize(number)
    raise ArgumentError unless (0..999_999_999_999).cover?(number)
    @number = number
  end

  def in_english(number = @number)
    case number
    when 0..19
      NUMBERS[number]
    when 20..99
      tens(number)
    when 100..999
      to_english(number, 100, 'hundred')
    when 1_000..999_999
      to_english(number, 1_000, 'thousand')
    when 1_000_000..999_999_999
      to_english(number, 1_000_000, 'million')
    when 1_000_000_000..999_999_999_999
      to_english(number, 1_000_000_000, 'billion')
    end
  end

  private

  def tens(number)
    string    = TENS[number / 10 - 2]
    remainder = number % 10

    string += "-#{in_english(remainder)}" unless remainder.zero?
    string
  end

  def to_english(number, decimal, decimal_unit)
    string    = "#{in_english(number / decimal)} #{decimal_unit}"
    remainder = number % decimal

    string += " #{in_english(remainder)}" unless remainder.zero?
    string
  end
end

File: go/say/say.go
package say

var numbers = []string{"zero", "one", "two", "three", "four", "five", "six",
	"seven", "eight", "nine", "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen",
	"sixteen", "seventeen", "eighteen", "nineteen"}

var tens = []string{"twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"}

func Say(input uint64) string {
	switch {
	case input < 20:
		return numbers[input]
	case input < 100:
		quotient, remainder := divmod(input, 10)
		result := tens[quotient-2]
		if remainder > 0 {
			result = result + "-" + Say(remainder)
		}
		return result
	case input < 1e3:
		return inEnglish(input, 1e2, "hundred")
	case input < 1e6:
		return inEnglish(input, 1e3, "thousand")
	case input < 1e9:
		return inEnglish(input, 1e6, "million")
	case input < 1e12:
		return inEnglish(input, 1e9, "billion")
	case input < 1e15:
		return inEnglish(input, 1e12, "trillion")
	case input < 1e18:
		return inEnglish(input, 1e15, "quadrillion")
	default:
		return inEnglish(input, 1e18, "quintillion")
	}
}

func inEnglish(number uint64, decimal uint64, decimalUnit string) string {
	quotient, remainder := divmod(number, decimal)
	result := Say(quotient) + " " + decimalUnit
	if remainder > 0 {
		result = result + " " + Say(remainder)
	}
	return result
}

func divmod(dividend uint64, divisor uint64) (uint64, uint64) {
	quotient := dividend / divisor
	remainder := dividend % divisor
	return quotient, remainder
}
